### ISSUE TITLE
Upgrade addEventListener method signatures for non-nullable types

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#9548]
The third parameter type is changed to boolean or dictionary.

```
[Exposed=(Window,Worker)]
interface EventTarget {
  void addEventListener(DOMString type, EventListener? callback, optional (AddEventListenerOptions or boolean) options);
  void removeEventListener(DOMString type, EventListener? callback, optional (EventListenerOptions or boolean) options);
  boolean dispatchEvent(Event event);
};

callback interface EventListener {
  void handleEvent(Event event);
};

dictionary EventListenerOptions {
  boolean capture = false;
};

dictionary AddEventListenerOptions : EventListenerOptions {
  boolean passive = false;
  boolean once = false;
};
```

https://dom.spec.whatwg.org/#interface-eventtarget

However, this is a large change. @zhengbli Could you upgrade using macros?

**Comment by appsforartists** [comment]
Meanwhile, to keep TypeScript from choking on your code until this is fixed upstream:

```typescript
interface WhatWGEventListenerArgs {
  capture?: boolean;
}

interface WhatWGAddEventListenerArgs extends WhatWGEventListenerArgs {
  passive?: boolean;
  once?: boolean;
}

type WhatWGAddEventListener = (
  type: string,
  listener: (event:Event) => void,
  options?: WhatWGAddEventListenerArgs
) => void;

// usage
(document.addEventListener as WhatWGAddEventListener)(
  'scroll',
  this._onScroll.bind(this),
  {
    passive: true,
  }
);
```

**Comment by reppners** [comment]
Has this been forgotten? Thought it would be contained in a current release but it isn't.

**Comment by mhegazy** [comment]
PRs welcomed. You can find more information about contributing lib.d.ts fixes at https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes.

**Comment by reppners** [comment]
Its fixed by https://github.com/Microsoft/TSJS-lib-generator/pull/224 👍

**Comment by epicgirl1998** [comment]
there are still many overrides for `addEventListener` remaining that define the third parameter as a boolean

you may do a search for `addEventListener` or `useCapture?: boolean` in [/src/lib/dom.generated.d.ts](https://github.com/Microsoft/TypeScript/blob/master/src/lib/dom.generated.d.ts) or [/src/lib/webworker.generated.d.ts](https://github.com/Microsoft/TypeScript/blob/master/src/lib/webworker.generated.d.ts)

they look like this:

```typescript
interface Document /* extends ... */ {
	/* ... */
    addEventListener<K extends keyof DocumentEventMap>(type: K, listener: (this: Document, ev: DocumentEventMap[K]) => any, useCapture?: boolean): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, useCapture?: boolean): void;
}
```

```typescript
// Argument of type '{ once: boolean; }' is not assignable to parameter of type 'boolean | undefined'.
//   Type '{ once: boolean; }' is not assignable to type 'false'.
document.addEventListener("DOMContentLoaded", function () {
	// ...
}, {once: true});
```

**Comment by dsifford** [comment]
@mhegazy Went to go take a stab at fixing this, but it looks like it was recently fixed in https://github.com/Microsoft/TSJS-lib-generator/commit/c0cb9d9147a50d74f8aaa0b806d82240af9d4ed9

The fix was targeting #18136, which appears to be a dupe of this issue.

Fix still hasn't made it to typescript stable yet though.

**Comment by mhegazy** [comment]
thanks @dsifford. this should be fixed now.
